README
 ----------------------------------------------------------------------
|                        JARSYNC - RSYNC IN JAVA                       |
 ----------------------------------------------------------------------

This is the source distribution of Jarsync, a library written in Java
that implements the rsync algorithm---a method for syncing remote files
by transmitting only the differences between the two files.

Jarsync is beta software, and is not thoroughly tested. 

=======================================================================

CONTENTS

1. About this library.
2. Compilation.
3. The API.
4. Files in this distribution.

=======================================================================

1. About this library.

The Rsync algorithm is a method for syncing two files on two different
computers connected via a low-bandwidth high-latency bidirectional
communications link. Instead of simply copying the new file onto the
old, it computes and sends checksums of the old file---which are,
altogether, possibly much smaller than the file itself---and the other
machine then uses to compute the differences between the old file and
the new. This machine then sends these differences back---again with
the hope that the differences are far smaller than the new file---and
the new file is recreated from these differences.

This library implements the rsync algorithm in Java, and provides a
simple API for using the algorithm.

For a more complete description of the rsync algorithm, see the paper
"The rsync algorithm", technical report TR-CS-96-05, by Andrew Tridgell
and Paul Mackerras, available from the URL

	http://rsync.samba.org/ftp/rsync/tech-report.ps

=======================================================================

2. Compilation.

To configure and build the library, the `rdiff' example program, and the
test programs, simply say

	./configure
	make

If this is a CVS checkout, first execute `autoconf' to make the
configure script.

There is also a Jakarta Ant build.xml file, which by executing

	ant
	ant test

will build the same files as `make'.

There is no `install' target yet; to use the test programs and the rdiff
workalike program the Jar file `jarsync.jar' needs to be in your CLASSPATH.

=======================================================================

3. The API.

The gist of the API (at this moment) is to provide the three most basic
functions in the rsync algorithm -- checksum generation, delta
generation, and file patching. A higher-level interface may be added
later.

This is only an overview; see the API JavaDocs for more information.

  3.a. Checksum Generation.

  Checksum generation is handled with the `Generator' class, and comes
  in two basic flavors: generating over byte arrays and generating over
  files. All generation methods are called `generateSums', which will
  all return a java.util.Collection of ChecksumPair objects.

  The constructor can take a `Configuration' argument, which stores
  things such as the message digest algorithm, the block size, etc.


  3.b. Delta Generation

  Deltas are generated with the `Matcher' class, whose methods take
  either a java.util.Collection of checksums or a TwoKeyMap generated
  from the `buildHashtable' method, and either a byte array, a
  java.io.File, or a java.io.InputStream. All methods are called
  `hashSearch' and all return a java.util.Collection of `Delta' objects,
  which are either of type `Offsets' or `DataBlock'.

  The Matcher class can take a `Configuration', similar to Generator.


  3.c. File Patching

  Files are patched with the `Rebuilder' class, whose methods
  `rebuildFile'  will either take a basis file and a java.util.Collection
  of Deltas and return a new, temporary file containing the patched
  file; or take two files and a java.util.Collection, patching into one
  of them. The `rebuildFileInPlace' will patch the file in-place, not
  creating any temporary files.

  Since these methods are stateless, they are all static and do not
  depend upon a `Configuration'.

  There are no equivalent InputStream/byte[] methods, as these methods
  require random access.

=======================================================================

4. Files in this distribution.

  AUTHORS      -  List of authors, roles, and contact information.
  COPYING      -  The GNU General Public License.
  ChangeLog    -  List of changes between versions.
  LICENSE      -  Decription of this library's license.
  Makefile.in  -  For producing Makefile with configure.
  README       -  This file; notes &c.
  TODO         -  A list of things yet to be done.
  build.xml    -  Jakarta Ant XML build file.
  configure    -  GNU configure script.
  configure.ac -  GNU autoconf specification.
  mutate.pl    -  Utility from librsync to make random modifications to files.
  rdiff        -  Bash script to run org.metastatic.rsync.Rdiff.
  rdiff.magic  -  Example (unofficial) magic(5) file for rdiff files.
  test.java    -  A simple test of byte array based methods.
  test2.java   -  A simple test of file-based methods.

In directory source/org/metastatic/rsync:

  BrokenMD4.java        -  A mutation of MD4 that aims to be compatible
                           with 
  ChecksumPair.java     -  A pair of weak and strong checksums over a
                           block of data.
  Configuration.java    -  A context object.
  DataBlock.java        -  A Delta that represents new data to be
                           inserted into a file.
  Delta.java            -  A generic change to data in the algorithm.
  Generator.java        -  The checksum generation methods.
  Matcher.java          -  The hash search and delta generation methods.
  MessageDigest.java    -  Superclass for message digest algorithms.
  MD4.java              -  The MD4 Message Digest algorithm, used to generate
                           the "strong" checksum.
  Offsets.java          -  A Delta that represents data that has simply
                           moved.
  Rdiff.java            -  rdiff(1)-workalike example program.
  Rebuilder.java        -  Reconstructs files from deltas.
  RollingChecksum.java  -  A simple checksum based on Adler32 that can
                           be updated from either end.
  TwoKeyMap.java        -  A java.util.Map that maps two keys, one an
                           integer and the other a byte array.

In directory getopt/gnu/getopt:

  Aaron M. Renn's Java port of GNU Getopt.

=======================================================================
  $Id: README,v 1.6 2002/08/23 19:57:21 rsdio Exp $


ChangeLog
ChangeLog for JARSYNC.

$Id: ChangeLog,v 1.2 2002/08/16 05:04:36 rsdio Exp $

2003-03-04 -- 0.1

  - The rsync-compatible program is on hold for now.

  - Beta release of the library. I think it's good enough for beta.

2002-08-15

  - Previous version numbers are now obsolete; right now there is no
    public release of the new source tree.

2002-08-14 -- 0.0.7

  - Not dead!

  - Administrata changes: all the classes are now in the
    org.metastatic.rsync package; MD4 is just a lone class and is not
    a JCE implementation; seperation of generation and matching methods
    into thier own respective classes; renamed 'Update' to 'Delta' --
    sounds better.

  - Some methods are slightly more comprehensible.

  - RollingChecksum appears to work and appears compatible with rsync's
    weak sum.

  - Interest in a program/library such as this appears to be on the
    rise; time to get going with developing this.

  - Moved to GPL.

2002-04-24 -- 0.0.6

  - Simple speed improvements to MD4 (inlining instead of calling the
    three functions FF, GG, and HH).

  - The rolling checksum made probably faster, better. I was using two
    long integers to store the two 16-bit parts of the checksum, not
    realizing that since Java's integral arithmetic is based on two's
    complement representations, that addition and subtraction act as
    though the numbers are unsigned. Thus two short integers instead
    (each half of the checksum is 16 bits) of long integers and doing
    all arithmetic explicitly modulo 2^32.

  - Removed logging statements from parts of Rsync.generateUpdates, to
    see how fast it runs for real. I could get about 43kb/sec on a
    Pentium 450 running Linux. This is not speedy, but hopefully given a
    modern machine it is good enough.

2002-04-08 -- 0.0.5

  - Core algorithm methods in place, except for memory-mapped File
    versions (if they are needed; it might be better to limit ourselves
    to 1.2 in the final release (using 1.4 is ok for development, within
    reason -- ie logging)). Tests over these methods appear to succeed.

  - TwoKeyMap now a full, if strange, Collection.

  - Checksums look almost compatible with the C implementation.

  - Class org.metastatic.util.Debug is obsoleted. Using
    java.util.logging classes for debug output now.


source/org/metastatic/rsync/MD4.java
/* vim:set softtabstop=3 shiftwidth=3 tabstop=3 expandtab tw=72:
   $Id: MD4.java,v 1.9 2003/03/30 15:18:46 rsdio Exp $
  
   This version is derived from the version in GNU Crypto.
  
   MD4: The MD4 message digest algorithm.
   Copyright (C) 2002 The Free Software Foundation, Inc.
   Copyright (C) 2003  Casey Marshall <rsdio@metastatic.org>
  
   This file is a part of Jarsync.
  
   Jarsync is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2 of the License, or (at
   your option) any later version.
  
   Jarsync is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with Jarsync; if not, write to the
  
      Free Software Foundation, Inc.,
      59 Temple Place, Suite 330,
      Boston, MA  02111-1307
      USA
  
   Linking Jarsync statically or dynamically with other modules is
   making a combined work based on Jarsync.  Thus, the terms and
   conditions of the GNU General Public License cover the whole
   combination.
  
   As a special exception, the copyright holders of Jarsync give you
   permission to link Jarsync with independent modules to produce an
   executable, regardless of the license terms of these independent
   modules, and to copy and distribute the resulting executable under
   terms of your choice, provided that you also meet, for each linked
   independent module, the terms and conditions of the license of that
   module.  An independent module is a module which is not derived from
   or based on Jarsync.  If you modify Jarsync, you may extend this
   exception to your version of it, but you are not obligated to do so.
   If you do not wish to do so, delete this exception statement from
   your version.  */

package org.metastatic.rsync;

import java.security.DigestException;
import java.security.MessageDigestSpi;

/**
 * <p>An implementation of Ron Rivest's MD4 message digest algorithm.
 * MD4 was the precursor to the stronger MD5
 * algorithm, and while not considered cryptograpically secure itself,
 * MD4 is in use in various applications. It is slightly faster than
 * MD5.</p>
 *
 * <p>This implementation is derived from the version of MD4 in <a
 * href="http://www.gnu.org/software/gnu-crypto/">GNU Crypto.</p>
 *
 * <p>References:</p>
 *
 * <ol>
 *    <li>The <a href="http://www.ietf.org/rfc/rfc1320.txt">MD4</a> Message-
 *    Digest Algorithm.<br>
 *    R. Rivest.</li>
 * </ol>
 *
 * @version $Revision: 1.9 $
 */
public class MD4 extends MessageDigestSpi implements Cloneable {

   // Constants and variables.
   // -----------------------------------------------------------------
 
   /**
    * An MD4 message digest is always 128-bits long, or 16 bytes.
    */
   public static final int DIGEST_LENGTH = 16;

   /**
    * The MD4 algorithm operates on 512-bit blocks, or 64 bytes.
    */
   public static final int BLOCK_LENGTH = 64;

   protected static final int A = 0x67452301;
   protected static final int B = 0xefcdab89;
   protected static final int C = 0x98badcfe;
   protected static final int D = 0x10325476;

   /* The four chaining variables. */
   protected int a, b, c, d;

   protected long count;

   protected final byte[] buffer;

   /** Word buffer for transforming. */
   private final int[] X = new int[16];

 // Constructors.
   // -----------------------------------------------------------------

   /**
    * Trivial zero-argument constructor.
    */
   public MD4() {
      buffer = new byte[BLOCK_LENGTH];
      engineReset();
   }

   /**
    * Private constructor for cloning.
    */
   private MD4(MD4 that) {
      this();

      this.a = that.a;
      this.b = that.b;
      this.c = that.c;
      this.d = that.d;
      this.count = that.count;
      System.arraycopy(that.buffer, 0, this.buffer, 0, BLOCK_LENGTH);
   }

   // java.lang.Cloneable interface implementation --------------------

   public Object clone() {
      return new MD4(this);
   }

   // SPI instance methods.
   // -----------------------------------------------------------------

   protected int engineGetDigestLength() {
      return DIGEST_LENGTH;
   }

   public void engineUpdate(byte b) {
      // compute number of bytes still unhashed; ie. present in buffer
      int i = (int)(count % BLOCK_LENGTH);
      count++;
      buffer[i] = b;
      if (i == (BLOCK_LENGTH - 1)) {
         transform(buffer, 0);
      }
   }

   protected void engineUpdate(byte[] b, int offset, int len) {
      int n = (int)(count % BLOCK_LENGTH);
      count += len;
      int partLen = BLOCK_LENGTH - n;
      int i = 0;

      if (len >= partLen) {
         System.arraycopy(b, offset, buffer, n, partLen);
         transform(buffer, 0);
         for (i = partLen; i + BLOCK_LENGTH - 1 < len; i+= BLOCK_LENGTH) {
            transform(b, offset + i);
         }
         n = 0;
      }

      if (i < len) {
         System.arraycopy(b, offset + i, buffer, n, len - i);
      }
   }

   /**
    * Pack the four chaining variables into a byte array.
    */
   protected byte[] engineDigest() {
      byte[] tail = padBuffer();
      engineUpdate(tail, 0, tail.length);
      byte[] digest = {
         (byte) a, (byte) (a >>> 8), (byte) (a >>> 16), (byte) (a >>> 24),
         (byte) b, (byte) (b >>> 8), (byte) (b >>> 16), (byte) (b >>> 24),
         (byte) c, (byte) (c >>> 8), (byte) (c >>> 16), (byte) (c >>> 24),
         (byte) d, (byte) (d >>> 8), (byte) (d >>> 16), (byte) (d >>> 24)
      };

      engineReset();

      return digest;
   }

   protected
   int engineDigest(byte[] out, int off, int len) throws DigestException {
      if (off < 0 || off + len >= out.length) {
         throw new DigestException();
      }
      System.arraycopy(engineDigest(), 0, out, off,
         Math.min(len, DIGEST_LENGTH));
      return Math.min(len, DIGEST_LENGTH);
   }

   /** Reset the four chaining variables. */
   protected void engineReset() {
      a = A; b = B;
      c = C; d = D;
      count = 0;
   }

   /**
    * Pad the buffer by appending the byte 0x80, then as many zero bytes
    * to fill the buffer 8 bytes shy of being a multiple of 64 bytes, then
    * append the length of the buffer, in bits, before padding.
    */
   protected byte[] padBuffer() {
      int n = (int) (count % BLOCK_LENGTH);
      int padding = (n < 56) ? (56 - n) : (120 - n);
      byte[] pad = new byte[padding + 8];

      pad[0] = (byte) 0x80;
      long bits = count << 3;
      pad[padding++] = (byte)  bits;
      pad[padding++] = (byte) (bits >>>  8);
      pad[padding++] = (byte) (bits >>> 16);
      pad[padding++] = (byte) (bits >>> 24);
      pad[padding++] = (byte) (bits >>> 32);
      pad[padding++] = (byte) (bits >>> 40);
      pad[padding++] = (byte) (bits >>> 48);
      pad[padding  ] = (byte) (bits >>> 56);

      return pad;
   }

   /** Transform a 64-byte block. */
   protected void transform(byte[] in, int offset) {
      int aa, bb, cc, dd;

      for (int i = 0, n = 0; i < 16; i++) {
         X[i] = (in[offset++] & 0xff)       |
                (in[offset++] & 0xff) <<  8 |
                (in[offset++] & 0xff) << 16 |
                (in[offset++] & 0xff) << 24;
      }

      aa = a;  bb = b;  cc = c;  dd = d;

      // Round 1
      a += ((b & c) | ((~b) & d)) + X[ 0];
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & b) | ((~a) & c)) + X[ 1];
      d = d <<  7 | d >>> (32 -  7);
      c += ((d & a) | ((~d) & b)) + X[ 2];
      c = c << 11 | c >>> (32 - 11);
      b += ((c & d) | ((~c) & a)) + X[ 3];
      b = b << 19 | b >>> (32 - 19);
      a += ((b & c) | ((~b) & d)) + X[ 4];
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & b) | ((~a) & c)) + X[ 5];
      d = d <<  7 | d >>> (32 -  7);
      c += ((d & a) | ((~d) & b)) + X[ 6];
      c = c << 11 | c >>> (32 - 11);
      b += ((c & d) | ((~c) & a)) + X[ 7];
      b = b << 19 | b >>> (32 - 19);
      a += ((b & c) | ((~b) & d)) + X[ 8];
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & b) | ((~a) & c)) + X[ 9];
      d = d <<  7 | d >>> (32 -  7);
      c += ((d & a) | ((~d) & b)) + X[10];
      c = c << 11 | c >>> (32 - 11);
      b += ((c & d) | ((~c) & a)) + X[11];
      b = b << 19 | b >>> (32 - 19);
      a += ((b & c) | ((~b) & d)) + X[12];
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & b) | ((~a) & c)) + X[13];
      d = d <<  7 | d >>> (32 -  7);
      c += ((d & a) | ((~d) & b)) + X[14];
      c = c << 11 | c >>> (32 - 11);
      b += ((c & d) | ((~c) & a)) + X[15];
      b = b << 19 | b >>> (32 - 19);

      // Round 2.
      a += ((b & (c | d)) | (c & d)) + X[ 0] + 0x5a827999;
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & (b | c)) | (b & c)) + X[ 4] + 0x5a827999;
      d = d <<  5 | d >>> (32 -  5);
      c += ((d & (a | b)) | (a & b)) + X[ 8] + 0x5a827999;
      c = c <<  9 | c >>> (32 -  9);
      b += ((c & (d | a)) | (d & a)) + X[12] + 0x5a827999;
      b = b << 13 | b >>> (32 - 13);
      a += ((b & (c | d)) | (c & d)) + X[ 1] + 0x5a827999;
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & (b | c)) | (b & c)) + X[ 5] + 0x5a827999;
      d = d <<  5 | d >>> (32 -  5);
      c += ((d & (a | b)) | (a & b)) + X[ 9] + 0x5a827999;
      c = c <<  9 | c >>> (32 -  9);
      b += ((c & (d | a)) | (d & a)) + X[13] + 0x5a827999;
      b = b << 13 | b >>> (32 - 13);
      a += ((b & (c | d)) | (c & d)) + X[ 2] + 0x5a827999;
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & (b | c)) | (b & c)) + X[ 6] + 0x5a827999;
      d = d <<  5 | d >>> (32 -  5);
      c += ((d & (a | b)) | (a & b)) + X[10] + 0x5a827999;
      c = c <<  9 | c >>> (32 -  9);
      b += ((c & (d | a)) | (d & a)) + X[14] + 0x5a827999;
      b = b << 13 | b >>> (32 - 13);
      a += ((b & (c | d)) | (c & d)) + X[ 3] + 0x5a827999;
      a = a <<  3 | a >>> (32 -  3);
      d += ((a & (b | c)) | (b & c)) + X[ 7] + 0x5a827999;
      d = d <<  5 | d >>> (32 -  5);
      c += ((d & (a | b)) | (a & b)) + X[11] + 0x5a827999;
      c = c <<  9 | c >>> (32 -  9);
      b += ((c & (d | a)) | (d & a)) + X[15] + 0x5a827999;
      b = b << 13 | b >>> (32 - 13);

      // Round 3.
      a += (b ^ c ^ d) + X[ 0] + 0x6ed9eba1;
      a = a <<  3 | a >>> (32 -  3);
      d += (a ^ b ^ c) + X[ 8] + 0x6ed9eba1;
      d = d <<  9 | d >>> (32 -  9);
      c += (d ^ a ^ b) + X[ 4] + 0x6ed9eba1;
      c = c << 11 | c >>> (32 - 11);
      b += (c ^ d ^ a) + X[12] + 0x6ed9eba1;
      b = b << 15 | b >>> (32 - 15);
      a += (b ^ c ^ d) + X[ 2] + 0x6ed9eba1;
      a = a <<  3 | a >>> (32 -  3);
      d += (a ^ b ^ c) + X[10] + 0x6ed9eba1;
      d = d <<  9 | d >>> (32 -  9);
      c += (d ^ a ^ b) + X[ 6] + 0x6ed9eba1;
      c = c << 11 | c >>> (32 - 11);
      b += (c ^ d ^ a) + X[14] + 0x6ed9eba1;
      b = b << 15 | b >>> (32 - 15);
      a += (b ^ c ^ d) + X[ 1] + 0x6ed9eba1;
      a = a <<  3 | a >>> (32 -  3);
      d += (a ^ b ^ c) + X[ 9] + 0x6ed9eba1;
      d = d <<  9 | d >>> (32 -  9);
      c += (d ^ a ^ b) + X[ 5] + 0x6ed9eba1;
      c = c << 11 | c >>> (32 - 11);
      b += (c ^ d ^ a) + X[13] + 0x6ed9eba1;
      b = b << 15 | b >>> (32 - 15);
      a += (b ^ c ^ d) + X[ 3] + 0x6ed9eba1;
      a = a <<  3 | a >>> (32 -  3);
      d += (a ^ b ^ c) + X[11] + 0x6ed9eba1;
      d = d <<  9 | d >>> (32 -  9);
      c += (d ^ a ^ b) + X[ 7] + 0x6ed9eba1;
      c = c << 11 | c >>> (32 - 11);
      b += (c ^ d ^ a) + X[15] + 0x6ed9eba1;
      b = b << 15 | b >>> (32 - 15);

      a += aa; b += bb; c += cc; d += dd;
   }

} 
